### PR TITLE
ExecutorManager refactor

### PR DESCRIPTION
**Pull request by chengren311**
1. Simplify logic
2. Change activeExecutors from HashSet to ImmutableSet to guarantee thread safety.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 1]
```diff
@@ -232,19 +234,19 @@ public void setupExecutors() throws ExecutorManagerException {
       }
       newExecutors.add(new Executor(executor.getId(), executorHost,
           executorPort, true));
+    } else {
+      //todo chengren311: convert to slf4j and parameterized logging
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: May we add a comment here to explain what cases should go into the else block?

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 2]
```diff
@@ -105,6 +105,8 @@
   private final int maxConcurrentRunsOneFlow;
   QueuedExecutions queuedFlows;
   File cacheDir;
+  //make it immutable to ensure threadsafety
+  private volatile ImmutableSet<Executor> activeExecutors = null;
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: How is the thread-safe guaranteed compared to HashSet?

Comment (chengren311) [reply azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: since immutable data structure is always thread safe.

Comment (kunkun-tang) [reply azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: I'd like to learn which part of code need the thread-safe data structure, activeExecutors? Would you mind point it out?

Comment (chengren311) [reply azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: yes, so the class contains multiple pieces iterating over activeExecutor, and code piece populating activeExecutors from database. activeExecutors should be immune to race condition when iterating and populating activeExecutors are in different thread context.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 3]
```diff
@@ -256,41 +260,39 @@ private boolean isMultiExecutorMode() {
...
-        } catch (final TimeoutException e) {
-          wasSuccess = false;
-          logger.error("Timed out while waiting for ExecutorInfo refresh"
-              + executor, e);
-        } catch (final Exception e) {
-          wasSuccess = false;
-          logger.error("Failed to update ExecutorInfo for executor : "
-              + executor, e);
-        }
+    final List<Pair<Executor, Future<ExecutorInfo>>> futures =
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: this just removes synchronized (this.activeExecutors)

**Pull request merged**